File: BricksFallingWhenHit/BricksFallingWhenHit.py
from typing import List

"""
This is the solution for the problem:
"Bricks Falling When Hit" (LeetCode #803, Hard)

Link:
https://leetcode.com/problems/bricks-falling-when-hit/

Problem summary:
- Given a grid of bricks (1 = brick, 0 = empty).
- A brick is stable if:
    - it is in the top row, OR
    - connected (4-directionally) to another stable brick.
- Each hit removes one brick.
- After each removal, unstable bricks fall.
- Return how many bricks fall after each hit.

Core idea (Reverse Flood Fill):
- Removing bricks dynamically is expensive.
- Instead:
    1) Apply all hits first.
    2) Flood-fill from the roof to mark stable bricks.
    3) Add bricks back in reverse order.
    4) When a brick reconnects to stability, flood-fill its component.
- Each brick is marked stable at most once → linear complexity.
"""


class Solution:
    def hitBricks(self, grid: List[List[int]], hits: List[List[int]]) -> List[int]:
        m, n = len(grid), len(grid[0])

        # -------------------------------------------------------
        # Flood fill: marks stable bricks with value 2
        # Returns size of connected stable component
        # -------------------------------------------------------
        def flood_fill(r: int, c: int) -> int:
            if r < 0 or r >= m or c < 0 or c >= n or grid[r][c] != 1:
                return 0

            grid[r][c] = 2  # mark as stable

            return (
                1
                + flood_fill(r + 1, c)
                + flood_fill(r - 1, c)
                + flood_fill(r, c + 1)
                + flood_fill(r, c - 1)
            )

        # -------------------------------------------------------
        # Check if a brick connects to already stable structure
        # -------------------------------------------------------
        def is_connected(r: int, c: int) -> bool:
            if r == 0:
                return True

            directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if 0 <= nr < m and 0 <= nc < n and grid[nr][nc] == 2:
                    return True

            return False

        # -------------------------------------------------------
        # Step 1: Remove all hits
        # -------------------------------------------------------
        for r, c in hits:
            grid[r][c] -= 1

        # -------------------------------------------------------
        # Step 2: Mark initially stable bricks (roof-connected)
        # -------------------------------------------------------
        for c in range(n):
            flood_fill(0, c)

        # -------------------------------------------------------
        # Step 3: Process hits in reverse
        # -------------------------------------------------------
        result = [0] * len(hits)

        for k in range(len(hits) - 1, -1, -1):
            r, c = hits[k]

            # restore brick
            grid[r][c] += 1

            # if brick exists and connects to stable area
            if grid[r][c] == 1 and is_connected(r, c):
                # minus 1 excludes the restored brick itself
                result[k] = flood_fill(r, c) - 1

        return result


# -------------------------------
# Driver code & test harness
# -------------------------------

def run_test(grid, hits, expected):
    solution = Solution()
    result = solution.hitBricks([row[:] for row in grid], hits)

    print(f"grid = {grid}")
    print(f"hits = {hits}")
    print(f"Expected: {expected}")
    print(f"Result:   {result}")
    print("✓ PASS" if result == expected else "✗ FAIL")
    print("-" * 60)


if __name__ == "__main__":
    print("=== Testing Bricks Falling When Hit (Reverse Flood Fill) ===\n")

    # Test Case 1: LeetCode example
    run_test(
        grid=[[1, 0, 0, 0], [1, 1, 1, 0]],
        hits=[[1, 0]],
        expected=[2]
    )

    # Test Case 2: Multiple hits
    run_test(
        grid=[[1,0,0,0],[1,1,0,0]],
        hits=[[1,1],[1,0]],
        expected=[0,0]
    )

    # Test Case 3: Top-row stability
    run_test(
        grid=[[1]],
        hits=[[0,0]],
        expected=[0]
    )

    # Test Case 4: Complex structure
    run_test(
        grid=[[1,1,1],[0,1,0],[0,0,0]],
        hits=[[0,2],[2,0],[0,1],[1,2]],
        expected=[0,0,1,0]
    )

    print("\n=== Algorithm Analysis ===\n")
    print("Time Complexity:")
    print("  - Each brick is flood-filled at most once")
    print("  - Total: O(m * n + k)")
    print("\nSpace Complexity:")
    print("  - O(m * n) worst-case recursion stack")
    print("\nKey Insights:")
    print("  - Reverse processing avoids repeated recomputation")
    print("  - Stable bricks are explicitly marked (value = 2)")
    print("  - Flood-fill acts as incremental connectivity expansion")

File: BricksFallingWhenHit/BricksFallingWhenHit.cpp
#include <iostream>
#include <vector>
#include <array>

using namespace std;

/*
This is the solution for the problem:
"Bricks Falling When Hit" (LeetCode #803, Hard)

Link:
https://leetcode.com/problems/bricks-falling-when-hit/

Problem summary:
- Given a grid of bricks (1 = brick, 0 = empty).
- A brick is stable if:
    - it is in the top row, OR
    - connected (4-directionally) to another stable brick.
- Each hit removes one brick.
- After each removal, unstable bricks fall.
- Return how many bricks fall after each hit.

Core idea (Reverse Flood Fill):
- Apply all hits first.
- Flood-fill from the top row and mark stable bricks as value 2.
- Process hits in reverse order:
    - restore brick
    - if connected to stability, flood-fill and count new stable bricks.
- Each brick becomes stable at most once → O(m*n + k).
*/

class Solution {
private:
    array<pair<int,int>,4> dir{{{-1,0},{1,0},{0,1},{0,-1}}};
    int m, n;

    bool valid(int x, int y) const noexcept {
        return x >= 0 && x < m && y >= 0 && y < n;
    }

    // Flood-fill stable bricks and return component size
    int dfs(vector<vector<int>>& grid, int x, int y) {
        if (grid[x][y] == 2) return 0;

        grid[x][y] = 2;
        int cnt = 1;

        for (const auto& [dx, dy] : dir) {
            int nx = x + dx;
            int ny = y + dy;
            if (valid(nx, ny) && grid[nx][ny] == 1) {
                cnt += dfs(grid, nx, ny);
            }
        }
        return cnt;
    }

    // Check whether this brick connects to stable structure
    bool is_connected(vector<vector<int>>& grid, int x, int y) noexcept {
        if (x == 0) return true;

        for (const auto& [dx, dy] : dir) {
            int nx = x + dx;
            int ny = y + dy;
            if (valid(nx, ny) && grid[nx][ny] == 2) {
                return true;
            }
        }
        return false;
    }

public:
    vector<int> hitBricks(vector<vector<int>>& grid,
                          vector<vector<int>>& hits) {

        m = grid.size();
        n = grid[0].size();

        // Step 1: Remove all hits
        for (const auto& h : hits) {
            grid[h[0]][h[1]] -= 1;
        }

        // Step 2: Mark stable bricks from the top row
        for (int j = 0; j < n; ++j) {
            if (grid[0][j] == 1) {
                dfs(grid, 0, j);
            }
        }

        // Step 3: Process hits in reverse
        vector<int> res(hits.size(), 0);

        for (int i = (int)hits.size() - 1; i >= 0; --i) {
            int x = hits[i][0];
            int y = hits[i][1];

            // restore brick
            grid[x][y] += 1;

            if (grid[x][y] == 1 && is_connected(grid, x, y)) {
                // minus restored brick itself
                res[i] = dfs(grid, x, y) - 1;
            }
        }

        return res;
    }
};


// ------------------------------------------------
// Driver code & test harness
// ------------------------------------------------

void run_test(vector<vector<int>> grid,
              vector<vector<int>> hits,
              vector<int> expected) {

    Solution solution;
    vector<int> result = solution.hitBricks(grid, hits);

    cout << "grid = [";
    for (const auto& row : grid) {
        cout << "[";
        for (int v : row) cout << v << " ";
        cout << "]";
    }
    cout << "]\n";

    cout << "hits = [";
    for (const auto& h : hits) {
        cout << "[" << h[0] << "," << h[1] << "]";
    }
    cout << "]\n";

    cout << "Expected: ";
    for (int x : expected) cout << x << " ";
    cout << "\nResult:   ";
    for (int x : result) cout << x << " ";

    bool pass = (result == expected);
    cout << "\n" << (pass ? "✓ PASS" : "✗ FAIL") << "\n";
    cout << "--------------------------------------------------\n";
}


int main() {
    cout << "=== Testing Bricks Falling When Hit (Reverse Flood Fill) ===\n\n";

    // Test Case 1: LeetCode example
    run_test(
        {{1,0,0,0},{1,1,1,0}},
        {{1,0}},
        {2}
    );

    // Test Case 2
    run_test(
        {{1,0,0,0},{1,1,0,0}},
        {{1,1},{1,0}},
        {0,0}
    );

    // Test Case 3: single brick
    run_test(
        {{1}},
        {{0,0}},
        {0}
    );

    // Test Case 4
    run_test(
        {{1,1,1},{0,1,0},{0,0,0}},
        {{0,2},{2,0},{0,1},{1,2}},
        {0,0,1,0}
    );

    cout << "\n=== Algorithm Analysis ===\n\n";
    cout << "Time Complexity:\n";
    cout << "  - Each brick is flood-filled at most once\n";
    cout << "  - Total: O(m * n + k)\n\n";
    cout << "Space Complexity:\n";
    cout << "  - O(m * n) recursion stack (worst case)\n\n";
    cout << "Key Insights:\n";
    cout << "  - Reverse processing avoids repeated recomputation\n";
    cout << "  - Stable bricks are explicitly marked (value = 2)\n";
    cout << "  - Flood-fill acts as incremental connectivity expansion\n";

    return 0;
}
